shaping/wrapping.go

```diff
@@ -0,0 +1,243 @@
+//     int out will yield the index of the first glyph corresponding to that rune.
+// lineStartRune - the index of the first rune in the line.
+// b - the line break candidate under consideration.
+// curLineWidth - the amount of space total in the current line.
+// curLineUsed - the amount of space in the current line that is already used.
+// nextLineWidth - the amount of space available on the next line.
+//
+// This function returns both a valid Output broken at b and a boolean
+// indicating whether the returned output should be used.
+func shouldKeepSegmentOnLine(out Output, runeToGlyph []int, lineStartRune int, b breakOption, curLineWidth, curLineUsed, nextLineWidth int) (candidateLine Output, keep bool) {
+	// Convert the break target to an inclusive index.
+	glyphStart, glyphEnd := inclusiveGlyphRange(lineStartRune, b.breakAtRune, runeToGlyph, len(out.Glyphs))
+
+	// Construct a line out of the inclusive glyph range.
+	candidateLine = out
+	candidateLine.Glyphs = candidateLine.Glyphs[glyphStart : glyphEnd+1]
+	candidateLine.RecomputeAdvance()
+	candidateAdvance := candidateLine.Advance.Ceil()
+	if candidateAdvance > curLineWidth && candidateAdvance-curLineUsed <= nextLineWidth {
+		// If it fits on the next line, put it there.
+		return candidateLine, false
+	}
+
+	return candidateLine, true
+}
+
+// Range indicates the location of a sequence of elements within a longer slice.
+type Range struct {
+	Count  int
+	Offset int
+}
+
+// lineWrap wraps the shaped glyphs of a paragraph to a particular max width.
+func lineWrap(out Output, dir di.Direction, paragraph []rune, runeToGlyph []int, breaks []breakOption, maxWidth int) []output {
+	var outputs []output
+	if len(breaks) == 0 {
+		// Pass empty lines through as empty.
+		outputs = append(outputs, output{
+			Shaped: out,
+			RuneRange: Range{
+				Count: len(paragraph),
+			},
+		})
+		return outputs
+	}
+
+	for i := 0; i < len(breaks); i++ {
+		b := breaks[i]
+		if b.breakAtRune+1 < len(runeToGlyph) {
+			// Check if this break is valid.
+			gIdx := runeToGlyph[b.breakAtRune]
+			g2Idx := runeToGlyph[b.breakAtRune+1]
+			cIdx := out.Glyphs[gIdx].ClusterIndex
+			c2Idx := out.Glyphs[g2Idx].ClusterIndex
+			if cIdx == c2Idx {
+				// This break is within a harfbuzz cluster, and is
+				// therefore invalid.
+				copy(breaks[i:], breaks[i+1:])
+				breaks = breaks[:len(breaks)-1]
+				i--
+			}
+		}
+	}
+
+	start := 0
+	runesProcessed := 0
+	for i := 0; i < len(breaks); i++ {
+		b := breaks[i]
+		// Always keep the first segment on a line.
+		good, _ := shouldKeepSegmentOnLine(out, runeToGlyph, start, b, maxWidth, 0, maxWidth)
+		end := b.breakAtRune
+	innerLoop:
+		for k := i + 1; k < len(breaks); k++ {
+			bb := breaks[k]
+			candidate, ok := shouldKeepSegmentOnLine(out, runeToGlyph, start, bb, maxWidth, good.Advance.Ceil(), maxWidth)
+			if ok {
+				// Use this new, longer segment.
+				good = candidate
+				end = bb.breakAtRune
+				i++
```

Comment:
Could you add a comment explaining why we go through ?

Reply:
done!